### PR TITLE
Added support for users to specify journals

### PR DESCRIPTION
**Pull request by ghost**
Previously this module assumed journals existed on the
same disk as the OSD. This patch allows for the user to
specify which journal they would like to use in their
hostname override file.

Closes-Bug: #1274649

### CODE REVIEW

**Comment by markvoelker** [comment]
So does this require a change at the composition layer in how you specify OSD disks?  E.g. would I need to change my h.1 puppet_openstack_builder YAML files once this change goes in?

**Comment by ghost** [comment]
it does, that change is going in upstream as we speak

**Comment by ghost** [comment]
Upstream change. Upstream is missing the has_compute lines. They will be added is a separate commit.

https://review.openstack.org/#/c/71140/

**Comment by markvoelker** [comment]
Ok, thanks for clarifying.  Given that this requires a change at the composition layer, I think we should consider adding in a backward compatibility layer.  In this patch we could do something like (note: untested code here, just trying to convey the goal):

$disk_hash = $name
$disk_array = split($disk_hash, ':')
$disk = $disk_array[0]
if $disk_array[1] {
  # Cool, the caller gave us what we now expect.
  $disk_journal = $disk_array[1]
} else {
  # So old school!  Tell the caller about the new format, but continue to Just Work
  warn "You appear to be using a deprecated OSD disk specification (you didn't specify a disk name:journal disk name pair).  Setting journal disk to $disk."
  $disk_journal = $disk
}

That way people using h.1 wouldn't see things go sidewise if they forget to change their YAML files when they go to h.2....they'd still get a deprecation warning so they know to make the change in the future, but the old behavior would be preserved.

File: manifests/osd.pp [hunk 1]
```diff
@@ -9,7 +9,39 @@
...
 
   include cephdeploy
-  $disk = $name
+
+  $disk_hash = $name
+  $disk_array = split($disk_hash, ':')
+  $disk = $disk_array[0]
+  $disk_journal = $disk_array[1]
+
+  if $disk_journal {
```
Comment (markvoelker) [inline manifests/osd.pp:17]: Do we actually need to duplicate the whole exec statement here?  Seems like maybe we could reduce the duplication a bit with something along the lines of:

```
$disk_hash = $name
$disk_array = split($disk_hash, ':')
$disk = $disk_array[0]
$disk_journal = $disk_array[1]
if $disk_journal {
  $disk_journal_real = $disk_journal
} else {
  $disk_journal_real = $disk
}
exec { "create osd $disk":
     cwd     => "/home/$user/bootstrap",
     command => "/usr/bin/ceph-deploy --overwrite-conf osd create $::hostname:$disk:$disk_journal_real",
     require => Exec["zap $disk"],
     unless  => "/usr/bin/test -e /home/$user/zapped/$disk",
  }
```

That way if no $disk_journal is specified you end up with the journal on the same disk, but we don't duplicate code for the entire exec.

File: manifests/osd.pp [hunk 2]
```diff
@@ -9,7 +9,39 @@
...
+
+    exec { "create osd $disk":
+      cwd     => "/home/$user/bootstrap",
+      command => "/usr/bin/ceph-deploy --overwrite-conf osd create $::hostname:$disk",
+      require => Exec["zap $disk"],
+      unless  => "/usr/bin/test -e /home/$user/zapped/$disk",
+    }
+
+  }
+  else {
```
Comment (markvoelker) [inline manifests/osd.pp]: I don't think this else statement can ever be executed as either the if or the elsif will always be executed instead.

Comment (ghost) [inline manifests/osd.pp]: Good point, I'll fix it up.

File: manifests/osd.pp (hunk 1 above)
Comment (ghost) [inline manifests/osd.pp:17]: I found that we do need both. Leaving that trailing colon on the end of the disk zap command results in the passing of an empty device to the command line (eg ceph-disk-zap host1 /dev/sdb /dev)

Comment (markvoelker) [inline manifests/osd.pp:17]: The method above doesn't leave a trailing colon....you wind up with:

/usr/bin/ceph-deploy --overwrite-conf osd create mynode:sda:sda

Comment (ghost) [inline manifests/osd.pp:17]: oh you wanted me to actually read your code....

**Comment by markvoelker** [comment]
This looks good.  +1 from me.  Might want to squash merge it to keep the history tidy though.

**Comment by rickerc** [comment]
+1

**Pull request merged**